src/libraries/Hooks.sol

```diff
@@ -233,21 +236,31 @@ library Hooks {
     /// @notice calls beforeSwap hook if permissioned and validates return value
     function beforeSwap(IHooks self, PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
         internal
-        returns (int256 amountToSwap, int128 hookDeltaSpecified)
+        returns (int256 amountToSwap, BeforeSwapDelta hookReturn)
     {
         amountToSwap = params.amountSpecified;
-        if (msg.sender == address(self)) return (amountToSwap, hookDeltaSpecified);
+        if (msg.sender == address(self)) return (amountToSwap, BeforeSwapDeltaLibrary.ZERO_DELTA);
+
         if (self.hasPermission(BEFORE_SWAP_FLAG)) {
-            hookDeltaSpecified = self.callHookWithReturnDelta(
-                abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
-                self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
-            ).toInt128();
+            bool canReturnDelta = self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG);
+            hookReturn = BeforeSwapDelta.wrap(
+                self.callHookWithReturnDelta(
+                    abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
+                    canReturnDelta
+                )
+            );
 
-            // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
-            if (hookDeltaSpecified != 0) {
-                bool exactInput = amountToSwap < 0;
-                amountToSwap += hookDeltaSpecified;
-                if (exactInput ? amountToSwap > 0 : amountToSwap < 0) revert HookDeltaExceedsSwapAmount();
+            // skip this logic for the case where the hook return is 0
+            if (canReturnDelta) {
+                // any return in unspecified is passed to the afterSwap hook for handling
+                int128 hookDeltaSpecified = hookReturn.getSpecifiedDelta();
```

Comment:
Can this logic be moved into `callHookWithReturnDelta` cause we check canReturnDelta twice now?

Reply:
Cannot do this as callHookWithReturnDelta is shared by different hooks

Reply:
hmm yeah, i tried playing around so we're not double checking canReturnDelta twice but nothing super clean :/